src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/PropertySetAnalysis/PropertySetAnalysis.PropertySetDataFlowOperationVisitor.cs

```diff
@@ -254,26 +257,33 @@ protected override PropertySetAbstractValue VisitAssignmentOperation(IAssignment
                     if (targetAnalysisEntity != null)
                     {
                         PointsToAbstractValue pointsToAbstractValue = this.GetPointsToAbstractValue(operation.Value);
-                        foreach (AbstractLocation abstractLocation in pointsToAbstractValue.Locations)
+                        if (!this.TrackedFieldPropertyAssignmentsOpt.TryGetValue(
+                                targetAnalysisEntity,
+                                out TrackedAssignmentData trackedAssignmentData))
                         {
-                            if (!this.TrackedFieldPropertyAssignmentsOpt.TryGetValue(
-                                    targetAnalysisEntity,
-                                    out PooledDictionary<AbstractLocation, PooledHashSet<IAssignmentOperation>> locationsToAssignments))
-                            {
-                                locationsToAssignments =
-                                    PooledDictionary<AbstractLocation, PooledHashSet<IAssignmentOperation>>.GetInstance();
-                                this.TrackedFieldPropertyAssignmentsOpt.Add(targetAnalysisEntity, locationsToAssignments);
-                            }
+                            trackedAssignmentData = new TrackedAssignmentData();
+                            this.TrackedFieldPropertyAssignmentsOpt.Add(targetAnalysisEntity, trackedAssignmentData);
+                        }
 
-                            if (!locationsToAssignments.TryGetValue(
-                                    abstractLocation,
-                                    out PooledHashSet<IAssignmentOperation> assignments))
+                        if (pointsToAbstractValue.Kind == PointsToAbstractValueKind.KnownLocations)
+                        {
+                            foreach (AbstractLocation abstractLocation in pointsToAbstractValue.Locations)
                             {
-                                assignments = PooledHashSet<IAssignmentOperation>.GetInstance();
-                                locationsToAssignments.Add(abstractLocation, assignments);
+                                trackedAssignmentData.TrackAssignmentWithAbstractLocation(operation, abstractLocation);
                             }
-
-                            assignments.Add(operation);
+                        }
+                        else if (pointsToAbstractValue.Kind == PointsToAbstractValueKind.Unknown
+                            || pointsToAbstractValue.Kind == PointsToAbstractValueKind.UnknownNotNull)
+                        {
+                            trackedAssignmentData.TrackAssignmentWithUnknownLocation(operation);
+                        }
+                        else if (pointsToAbstractValue.NullState == NullAbstractValue.Null)
+                        {
+                            // Do nothing.
+                        }
```

Comment:
I have a question about this condition. Although you do nothing when meet this condition, but I find that when "this.Settings = null" it goes into the first if statement. So, is there any condition that the NullState is Null but the Kind is not KnownLocations?

Reply:
I'm not sure.  @mavasani?

Reply:
Yes, this is possible for [PointsToAbstractValueKind.UnknownNull](https://github.com/dotnet/roslyn-analyzers/blob/5ed057905c4e4def7f9e5abd4baccf8fc7286b4a/src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/PointsToAnalysis/PointsToAbstractValueKind.cs#L36). For example, if you have an if check such as `if (this.field == null)`. We don't know the underlying locations that the field points to, but inside that if branch we know that it has null state and in the else branch it has non-null state but unknown locations. `UnknownNull` is basically a complementary part of `UnknownNotNull` as mentioned in their doc comments.